pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -32,6 +34,75 @@ import (
 // EndpointsMatch is a type of function that returns true if pod endpoints match.
 type EndpointsMatch func(*v1.Pod, *v1.Pod) bool
 
+// ServiceSelectorCache is a cache of service selectors to avoid high CPU consumption caused by frequent calls to AsSelectorPreValidated (see #73527)
+type ServiceSelectorCache struct {
+	cache map[string]labels.Selector
+	lock  sync.RWMutex
+}
+
+// NewServiceSelectorCache init ServiceSelectorCache for both endpoint controller and endpointSlice controller.
+func NewServiceSelectorCache() *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: map[string]labels.Selector{},
+	}
+}
+
+// Get return selector and existence in ServiceSelectorCache by key.
+func (sc *ServiceSelectorCache) Get(key string) (labels.Selector, bool) {
+	sc.lock.RLock()
+	selector, ok := sc.cache[key]
+	sc.lock.RUnlock()
```

Comment:
This seems like something that could be part of a defer block. Something like this:

```go
sc.lock.Rlock()
defer sc.lock.RUnlock()
return sc.cache[key]
```

Reply:
defer sentence is consistent with others. But I noticed that sc.lock.RUnlock() executed after selector, ok := sc.cache[key] will improve about 10% performance of BenchmarkGetPodServiceMemberships than defer sc.lock.RUnlock()

Reply:
:+1: sounds good, thanks for checking the performance difference!

Reply:
Requires at least a comment...

Reply:
Also, if you could share the results if benchmark for this - are we sure this isn't a noice?
Or you're saying that the memory allocation that defer is doing is contributing that much in this case?

Reply:
Addressed, thanks.